src/agent/src/policy.rs

```diff
@@ -0,0 +1,218 @@
+        // OPA are expected to work without retrying, after connecting
+        // successfully for the first time.
+        for i in 0..50 {
+            if i > 0 {
+                sleep(Duration::from_millis(100)).await;
+                println!("policy initialize: PUT failed, retrying");
+            }
+
+            // Set-up the default policy.
+            if let Ok(_) = self
+                .opa_client
+                .put(&self.policy_path)
+                .body(policy.clone())
+                .send()
+                .await
+            {
+                // Check if requests causing policy errors should actually
+                // be allowed. That is an insecure configuration but is
+                // useful for allowing insecure pods to start, then connect to
+                // them and inspect Guest logs for the root cause of a failure.
+                if let Ok(allow_failures) = self
+                    .post_query("AllowRequestsFailingPolicy", EMPTY_JSON_INPUT)
+                    .await
+                {
+                    self.allow_failures = allow_failures;
+                }
+                return Ok(());
+            }
+        }
+        Err(anyhow!("failed to connect to OPA"))
+    }
+
+    /// Ask OPA to check if an API call should be allowed or not.
+    pub async fn is_allowed_endpoint(&mut self, ep: &str, request: &str) -> bool {
+        let post_input = "{\"input\":".to_string() + request + "}";
+
+        #[cfg(feature = "agent-policy-debug")]
+        self.log_opa_input(ep, &post_input).await;
+
+        self.post_query(ep, &post_input).await.unwrap_or(false)
+    }
+
+    /// Replace the Policy in OPA.
+    pub async fn set_policy(&mut self, policy: &str) -> Result<()> {
+        // Delete the old rules.
+        self.opa_client
+            .delete(&self.policy_path)
+            .send()
+            .await
+            .map_err(|e| anyhow!(e))?;
+
+        // Put the new rules.
+        self.opa_client
+            .put(&self.policy_path)
+            .body(policy.to_string())
+            .send()
+            .await
+            .map_err(|e| anyhow!(e))?;
+
+        // Check if requests causing policy errors should actually be allowed.
+        // That is an insecure configuration but is useful for allowing insecure
+        // pods to start, then connect to them and inspect Guest logs for the
+        // root cause of a failure.
+        self.allow_failures = self
+            .post_query("AllowRequestsFailingPolicy", EMPTY_JSON_INPUT)
+            .await?;
+        Ok(())
+    }
+
+    // Post query to OPA.
+    async fn post_query(&mut self, ep: &str, post_input: &str) -> Result<bool> {
+        debug!(sl!(), "policy check: {}", ep);
+
+        let uri = self.query_path.clone() + ep;
+        let response = self
+            .opa_client
+            .post(uri)
+            .body(post_input.to_string())
+            .send()
+            .await
+            .map_err(|e| anyhow!(e))?;
```

Comment:
if we are using anyhow crate, we can directly use ? after await

Reply:
I will make this change.